pkg/innerring/processors/audit/process.go

```diff
@@ -0,0 +1,137 @@
+	"go.uber.org/zap"
+)
+
+const maxClientConnectionTries = 3 // todo: set as parameter
+
+var sgFilter object.SearchFilters
+
+func (ap *Processor) processStartAudit(epoch uint64) {
+	log := ap.log.With(zap.Uint64("epoch", epoch))
+
+	containers, err := ap.selectContainersToAudit(epoch)
+	if err != nil {
+		log.Error("container selection failure", zap.String("error", err.Error()))
+
+		return
+	}
+
+	log.Info("select containers for audit", zap.Int("amount", len(containers)))
+
+	nm, err := ap.netmapClient.GetNetMap(0)
+	if err != nil {
+		ap.log.Error("can't fetch network map",
+			zap.String("error", err.Error()))
+
+		return
+	}
+
+	for i := range containers {
+		cnr, err := ap.containerClient.Get(containers[i]) // get container structure
+		if err != nil {
+			log.Error("can't get container info, ignore",
+				zap.Stringer("cid", containers[i]),
+				zap.String("error", err.Error()))
+
+			continue
+		}
+
+		// find all container nodes for current epoch
+		nodes, err := nm.GetContainerNodes(cnr.PlacementPolicy(), nil)
+		if err != nil {
+			log.Info("can't build placement for container, ignore",
+				zap.Stringer("cid", containers[i]),
+				zap.String("error", err.Error()))
+
+			continue
+		}
+
+		// shuffle nodes to ask a random one
+		n := nodes.Flatten()
+		rand.Shuffle(len(n), func(i, j int) { // fixme: consider using crypto rand
+			n[i], n[j] = n[j], n[i]
+		})
+
+		// search storage groups
+		storageGroups := ap.findStorageGroups(containers[i], n)
+		log.Info("select storage groups for audit",
+			zap.Stringer("cid", containers[i]),
+			zap.Int("amount", len(storageGroups)))
+
+		// todo: for each container push audit tasks into queue
+	}
+}
+
+func (ap *Processor) findStorageGroups(cid *container.ID, shuffled netmap.Nodes) []*object.ID {
+	var sg []*object.ID
+
+	for i := range shuffled { // consider iterating over some part of container
+		log := ap.log.With(
+			zap.Stringer("cid", cid),
+			zap.String("address", shuffled[0].Address()),
+			zap.Int("try", i),
+			zap.Int("total_tries", maxClientConnectionTries),
+		)
+
+		addr, err := ipAddr(shuffled[i].Address())
+		if err != nil {
+			log.Warn("can't parse remote address", zap.String("error", err.Error()))
+		}
+
+		cli, err := ap.clientCache.Get(addr)
+		if err != nil {
+			log.Warn("can't setup remote connection", zap.String("error", err.Error()))
+
+			continue
+		}
+
+		sgSearchParams := &client.SearchObjectParams{}
+		sgSearchParams.WithContainerID(cid)
+		sgSearchParams.WithSearchFilters(storageGroupSearchFilter())
+
+		// fixme: timeout from config
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		result, err := cli.SearchObject(ctx, sgSearchParams)
+		cancel()
+
+		if err != nil {
+			log.Warn("error in storage group search", zap.String("error", err.Error()))
+			continue
+		}
+
+		sg = append(sg, result...)
+
+		break // we found storage groups, so break loop
+	}
+
+	return sg
+}
+
+func storageGroupSearchFilter() object.SearchFilters {
```

Comment:
Can we initialize immediately `var sgFilter = storagegroup.SearchQuery()`?

Reply:
Sure. Fixed.